libyul/backends/wasm/EVMToEWasmTranslator.cpp

```diff
@@ -122,32 +143,318 @@ function mul_128x128_256(x1, x2, y1, y2) -> r1, r2, r3, r4 {
 
 	r1 := i64.add(i64.add(ah, carry1), carry2)
 }
+// Multiplies two 256 bit values resulting in a 512 bit
+// value split into eight 64 bit values.
+function mul_256x256_512(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4, r5, r6, r7, r8 {
+	let a1, a2, a3, a4 := mul_128x128_256(x1, x2, y1, y2)
+	let         b1, b2, b3, b4 := mul_128x128_256(x1, x2, y3, y4)
+	let         c1, c2, c3, c4 := mul_128x128_256(x3, x4, y1, y2)
+	let                 d1, d2, d3, d4 := mul_128x128_256(x3, x4, y3, y4)
+
+	r8 := d4
+	r7 := d3
+
+	let carry1, carry2
+	let t1, t2
+
+	r6, carry1 := add_carry(b4, c4, 0)
+	r6, carry2 := add_carry(r6, d2, 0)
+
+	r5, carry1 := add_carry(b3, c3, carry1)
+	r5, carry2 := add_carry(r5, d1, carry2)
+
+	r4, carry1 := add_carry(a4, b2, carry1)
+	r4, carry2 := add_carry(r4, c2, carry2)
+
+	r3, carry1 := add_carry(a3, b1, carry1)
+	r3, carry2 := add_carry(r3, c1, carry2)
+
+	r2, carry1 := add_carry(a2, carry1, carry1)
+	r1 := i64.add(a1, carry1)
+}
 function mul(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
 	// TODO it would actually suffice to have mul_128x128_128 for the first two.
 	let b1, b2, b3, b4 := mul_128x128_256(x3, x4, y1, y2)
 	let c1, c2, c3, c4 := mul_128x128_256(x1, x2, y3, y4)
-	let         d1, d2, d3, d4 := mul_128x128_256(x3, x4, y3, y4)
+	let d1, d2, d3, d4 := mul_128x128_256(x3, x4, y3, y4)
 	r4 := d4
 	r3 := d3
 	let t1, t2
 	t1, t2, r1, r2 := add(0, 0, b3, b4, 0, 0, c3, c4)
 	t1, t2, r1, r2 := add(0, 0, r1, r2, 0, 0, d1, d2)
 }
 function div(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
-	// TODO implement properly
-	r4 := i64.div_u(x4, y4)
+	// Based on https://github.com/ewasm/evm2wasm/blob/master/wasm/DIV.wast
+	if iszero(y1, y2, y3, y4) {
+		invalid()
```

Comment:
Should return zero to be compliant with evm

Reply:
Done